File: tests/adapters/c/goldens/imports/keep_all.c
/**
 * C module for testing include optimization.
 */

// Standard library includes (external)
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include <stdbool.h>
#include <stddef.h>
#include <limits.h>
#include <float.h>
#include <math.h>
#include <time.h>
#include <errno.h>
#include <assert.h>

// POSIX headers (external/system)
#include <unistd.h>
#include <fcntl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <pthread.h>
#include <signal.h>
#include <dirent.h>

// Common external libraries (external)
#include <zlib.h>
#include <curl/curl.h>
#include <sqlite3.h>
#include <openssl/ssl.h>
#include <openssl/err.h>
#include <libpq-fe.h>
#include <json-c/json.h>
#include <pcre.h>

// Local/project includes (should be considered local)
#include "services/user_service.h"
#include "database/connection.h"
#include "errors/validation_error.h"
#include "errors/network_error.h"
#include "utils/helpers/date_formatter.h"
#include "utils/helpers/json_parser.h"
#include "types/api_response.h"
#include "types/user_model.h"
#include "types/post_model.h"

// Relative includes with different depth levels
#include "../shared/utilities.h"
#include "../../core/core_module.h"
#include "../../../config/app_config.h"

// Long include lists from single directory (candidates for summarization)
#include "validation/email_validator.h"
#include "validation/password_validator.h"
#include "validation/phone_validator.h"
#include "validation/postal_code_validator.h"
#include "validation/credit_card_validator.h"
#include "validation/input_sanitizer.h"
#include "validation/currency_formatter.h"
#include "validation/phone_formatter.h"
#include "validation/slug_generator.h"
#include "validation/hash_creator.h"
#include "validation/hash_verifier.h"

#include "operations/create_user.h"
#include "operations/update_user.h"
#include "operations/delete_user.h"
#include "operations/get_user_by_id.h"
#include "operations/get_user_by_email.h"
#include "operations/get_users_by_role.h"
#include "operations/get_users_with_pagination.h"
#include "operations/activate_user.h"
#include "operations/deactivate_user.h"
#include "operations/reset_user_password.h"
#include "operations/change_user_role.h"
#include "operations/validate_user_permissions.h"

typedef struct {
    void* user_service;
    void* db_connection;
    void* logger;
} ImportTestService;

ImportTestService* import_test_service_new(
    void* user_service,
    void* db_connection,
    void* logger
) {
    ImportTestService* service = (ImportTestService*)malloc(sizeof(ImportTestService));
    if (!service) return NULL;

    service->user_service = user_service;
    service->db_connection = db_connection;
    service->logger = logger;

    return service;
}

void* process_data(ImportTestService* service, void** data, int count) {
    if (!service || !data) return NULL;

    // Using external libraries (zlib for compression)
    z_stream stream;
    memset(&stream, 0, sizeof(stream));

    // Using standard library functions
    time_t now = time(NULL);
    struct tm* timeinfo = localtime(&now);

    // Using POSIX functions
    pid_t pid = getpid();

    // Using local utilities
    // (would call validation functions here)

    return NULL;
}

int make_http_request(ImportTestService* service, const char* url) {
    if (!service || !url) return -1;

    // Using libcurl
    CURL* curl = curl_easy_init();
    if (!curl) {
        fprintf(stderr, "Failed to initialize curl\n");
        return -1;
    }

    curl_easy_setopt(curl, CURLOPT_URL, url);
    curl_easy_setopt(curl, CURLOPT_TIMEOUT, 5L);
    curl_easy_setopt(curl, CURLOPT_USERAGENT, "ImportTestService/1.0");

    CURLcode res = curl_easy_perform(curl);
    curl_easy_cleanup(curl);

    if (res != CURLE_OK) {
        fprintf(stderr, "HTTP request failed: %s\n", curl_easy_strerror(res));
        return -1;
    }

    return 0;
}

char* serialize_data(void* data) {
    if (!data) return NULL;

    // Using json-c
    json_object* jobj = json_object_new_object();
    if (!jobj) return NULL;

    // Add data to JSON object
    json_object_object_add(jobj, "data", json_object_new_string("test"));

    const char* json_str = json_object_to_json_string(jobj);
    char* result = strdup(json_str);

    json_object_put(jobj);
    return result;
}

int query_database(ImportTestService* service, const char* sql) {
    if (!service || !sql) return -1;

    // Using libpq (PostgreSQL)
    PGconn* conn = NULL; // Would connect here
    if (!conn) {
        fprintf(stderr, "Database connection failed\n");
        return -1;
    }

    PGresult* res = PQexec(conn, sql);
    if (PQresultStatus(res) != PGRES_TUPLES_OK) {
        fprintf(stderr, "Query failed: %s\n", PQerrorMessage(conn));
        PQclear(res);
        return -1;
    }

    int rows = PQntuples(res);
    PQclear(res);

    return rows;
}

void free_import_test_service(ImportTestService* service) {
    if (!service) return;
    free(service);
}

File: tests/adapters/rust/goldens/imports/keep_all.rs
// Rust module for testing import optimization.

// Standard library imports (external)
use std::io::{self, Read, Write, BufReader, BufWriter};
use std::collections::{HashMap, HashSet, BTreeMap, BTreeSet};
use std::sync::{Arc, Mutex, RwLock};
use std::fs::{File, OpenOptions};
use std::path::{Path, PathBuf};
use std::time::{Duration, SystemTime, UNIX_EPOCH};
use std::fmt::{self, Display, Formatter};
use std::error::Error;
use std::convert::{TryFrom, TryInto};

// More standard library
use std::net::{TcpListener, TcpStream};
use std::thread;
use std::env;
use std::process::Command;

// Third-party library imports (external)
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use tokio::runtime::Runtime;
use tokio::sync::mpsc;
use reqwest::Client;
use chrono::{DateTime, Utc};
use log::{debug, error, info, warn};
use anyhow::{Result, Context as AnyhowContext};
use thiserror::Error as ThisError;

// Database crates
use sqlx::{PgPool, Row};
use diesel::prelude::*;
use redis::Commands;

// Web framework crates
use actix_web::{web, App, HttpServer, HttpResponse};
use axum::{Router, routing::get};
use warp::Filter;

// Local/project imports (should be considered local)
use crate::models::User;
use crate::services::UserService;
use crate::database::Connection;
use crate::errors::{ValidationError, NetworkError};
use crate::utils::helpers::{date_formatter, json_parser};
use crate::types::{ApiResponse, UserModel, PostModel};

// Relative imports with different depth levels
use super::utilities;
use super::super::core::core_module;
use super::super::super::config::app_config;

// Long import lists from single module (candidates for summarization)
use crate::validation::{
    email_validator,
    password_validator,
    phone_validator,
    postal_code_validator,
    credit_card_validator,
    input_sanitizer,
    currency_formatter,
    phone_formatter,
    slug_generator,
    hash_creator,
    hash_verifier,
};

use crate::operations::{
    create_user,
    update_user,
    delete_user,
    get_user_by_id,
    get_user_by_email,
    get_users_by_role,
    get_users_with_pagination,
    activate_user,
    deactivate_user,
    reset_user_password,
    change_user_role,
    validate_user_permissions,
};

struct ImportTestService {
    user_service: UserService,
    db_connection: Connection,
    logger: Box<dyn std::any::Any>,
}

impl ImportTestService {
    fn new(user_service: UserService, db_connection: Connection, logger: Box<dyn std::any::Any>) -> Self {
        Self {
            user_service,
            db_connection,
            logger,
        }
    }

    fn process_data(&self, data: Vec<Box<dyn std::any::Any>>) -> Vec<HashMap<String, String>> {
        // Using standard library
        let mut processed = Vec::new();

        for item in data {
            let mut result = HashMap::new();
            // Using chrono for timestamps
            let timestamp = Utc::now().to_rfc3339();
            result.insert("timestamp".to_string(), timestamp);
            processed.push(result);
        }

        // Using local utilities (would call validation functions here)
        processed
    }

    async fn make_http_request(&self, url: &str) -> Result<String> {
        // Using reqwest
        let client = Client::builder()
            .timeout(Duration::from_secs(5))
            .user_agent("ImportTestService/1.0")
            .build()?;

        let response = client.get(url).send().await?;

        if !response.status().is_success() {
            anyhow::bail!("HTTP request failed with status: {}", response.status());
        }

        let text = response.text().await?;
        Ok(text)
    }

    fn serialize_data(&self, data: &dyn std::any::Any) -> Result<String> {
        // Using serde_json
        let json_value = json!({
            "data": "test"
        });

        Ok(serde_json::to_string_pretty(&json_value)?)
    }

    async fn query_database(&self, sql: &str) -> Result<Vec<Row>> {
        // Using sqlx
        let pool = PgPool::connect("postgresql://localhost/test").await?;

        let rows = sqlx::query(sql).fetch_all(&pool).await?;

        Ok(rows)
    }

    async fn start_server(&self) -> Result<()> {
        // Using actix-web
        HttpServer::new(|| {
            App::new()
                .route("/health", web::get().to(|| async { HttpResponse::Ok().body("OK") }))
        })
        .bind(("127.0.0.1", 8080))?
        .run()
        .await?;

        Ok(())
    }
}

// Forward declarations (should not be treated as imports)
struct Forward;
mod module_forward;
